### PR TITLE
Part 2 changes for test dataset

### PR DESCRIPTION
**Pull request by FreddieNeverLeft**
# PR Type
**[Feature]**

# Short Description

[Clickup Ticket(s): Link](https://app.clickup.com/t/8685pr77r)
[Clickup Subticket: Link](https://app.clickup.com/t/8688fbyck)

As detailed in the ticket's subtasks, this branch will focus on subtask2:

Task 2: add test dataset with assumption that the num samples between val and test are different, we split the two metrics and aggregate them separately.

# Tests Added
updated test_metrics_reporter_evaluate() in test_basic_client.py, for test dataset metrics.

### CODE REVIEW

File: fl4health/clients/basic_client.py [hunk 1]
```diff
@@ -99,6 +99,7 @@ def __init__(
         self.test_loader: Optional[DataLoader]
         self.num_train_samples: int
         self.num_val_samples: int
+        self.num_test_samples: Optional[int]
```
Comment (emersodb) [inline fl4health/clients/basic_client.py:102]: I'd set this to None by default just to guarantee an expected value.

File: fl4health/clients/basic_client.py [hunk 2]
```diff
@@ -699,6 +700,8 @@ def validate(self) -> Tuple[float, Dict[str, Scalar]]:
                 self.test_loader, self.test_loss_meter, self.test_metric_manager, LoggingMode.TEST
             )
             # There will be no clashes due to the naming convention associated with the metric managers
+            if self.num_test_samples is not None:
+                val_metrics["test - num_examples"] = int(self.num_test_samples)
```
Comment (emersodb) [inline fl4health/clients/basic_client.py:704]: Is the `int` cast here necessary given that `self.num_test_samples` has type int after the if statement?

Comment (emersodb) [inline fl4health/clients/basic_client.py:704]: Perhaps we should formalize the "test -" prefix as an enum in utils and import it here and in the base server just to ensure that they don't get out of sync at some point later on?

File: fl4health/server/base_server.py [hunk 3]
```diff
@@ -163,6 +164,90 @@ def poll_clients_for_sample_counts(self, timeout: Optional[float]) -> List[int]:
...
+            timeout=timeout,
+        )
+        log(
+            DEBUG,
+            "evaluate_round %s received %s results and %s failures for Validation",
+            server_round,
+            len(results),
+            len(failures),
+        )
+        test_metric_found = any(
```
Comment (emersodb) [inline fl4health/server/base_server.py:224]: If we implement `_split_metrics()` in the right way, I don't think we need to do this key test. See above.

File: fl4health/server/base_server.py [hunk 4]
```diff
@@ -163,6 +164,90 @@ def poll_clients_for_sample_counts(self, timeout: Optional[float]) -> List[int]:
...
+            test_loss = float(test_metrics.pop("test - loss"))
+            test_num_examples = int(test_metrics.pop("test - num_examples"))
+
+            non_test_eval_res = EvaluateRes(eval_res.status, eval_res.loss, eval_res.num_examples, val_metrics)
+            test_eval_res = EvaluateRes(eval_res.status, test_loss, test_num_examples, test_metrics)
+
+            val_results.append((client_proxy, non_test_eval_res))
+            test_results.append((client_proxy, test_eval_res))
+
+        return val_results, test_results
```
Comment (emersodb) [inline fl4health/server/base_server.py:187]: I mocked up a slight change to this workflow with an assertion and avoiding the need to do any key checking in the `_evaluate_round` implementation (see below). Let me know if you see any issues with the potential changes.

Comment (emersodb) [reply fl4health/server/base_server.py:187]: `    def _unpack_metrics(
        self, results: List[Tuple[ClientProxy, EvaluateRes]]
    ) -> Tuple[List[Tuple[ClientProxy, EvaluateRes]], List[Tuple[ClientProxy, EvaluateRes]]]:
        val_results = []
        test_results = []

        for client_proxy, eval_res in results:
            val_metrics = {k: v for k, v in eval_res.metrics.items() if not k.startswith("test - ")}
            test_metrics = {k: v for k, v in eval_res.metrics.items() if k.startswith("test - ")}

            if len(test_metrics) > 0:
                assert "test - loss" in test_metrics and "test - num_examples" in test_metrics, (
                    "test - loss and test - num_examples keys must be present "
                    "in test_metrics dictionary for aggregation"
                )
                # Remove loss and num_examples from test_metrics if they exist
                test_loss = float(test_metrics.pop("test - loss"))
                test_num_examples = int(test_metrics.pop("test - num_examples"))
                test_eval_res = EvaluateRes(eval_res.status, test_loss, test_num_examples, test_metrics)
                test_results.append((client_proxy, test_eval_res))

            val_eval_res = EvaluateRes(eval_res.status, eval_res.loss, eval_res.num_examples, val_metrics)
            val_results.append((client_proxy, val_eval_res))

        return val_results, test_results`

Comment (emersodb) [reply fl4health/server/base_server.py:187]:
```
    def _unpack_metrics(
        self, results: List[Tuple[ClientProxy, EvaluateRes]]
    ) -> Tuple[List[Tuple[ClientProxy, EvaluateRes]], List[Tuple[ClientProxy, EvaluateRes]]]:
        val_results = []
        test_results = []

        for client_proxy, eval_res in results:
            val_metrics = {k: v for k, v in eval_res.metrics.items() if not k.startswith("test - ")}
            test_metrics = {k: v for k, v in eval_res.metrics.items() if k.startswith("test - ")}

            if len(test_metrics) > 0:
                assert "test - loss" in test_metrics and "test - num_examples" in test_metrics, (
                    "test - loss and test - num_examples keys must be present "
                    "in test_metrics dictionary for aggregation"
                )
                # Remove loss and num_examples from test_metrics if they exist
                test_loss = float(test_metrics.pop("test - loss"))
                test_num_examples = int(test_metrics.pop("test - num_examples"))
                test_eval_res = EvaluateRes(eval_res.status, test_loss, test_num_examples, test_metrics)
                test_results.append((client_proxy, test_eval_res))

            val_eval_res = EvaluateRes(eval_res.status, eval_res.loss, eval_res.num_examples, val_metrics)
            val_results.append((client_proxy, val_eval_res))

        return val_results, test_results
```

File: fl4health/server/base_server.py (hunk 3 above)
Comment (emersodb) [inline fl4health/server/base_server.py:224]: I mocked up a slight simplification using the code for the metric unpacking above for this workflow. Let me know what you think 
```
        val_results, test_results = self._unpack_metrics(results)

        # Aggregate the validation results
        val_aggregated_result: Tuple[
            Optional[float],
            Dict[str, Scalar],
        ] = self.strategy.aggregate_evaluate(server_round, val_results, failures)
        val_loss_aggregated, val_metrics_aggregated = val_aggregated_result

        # Aggregate the test results if they are present
        if len(test_results) > 0:
            test_aggregated_result: Tuple[
                Optional[float],
                Dict[str, Scalar],
            ] = self.strategy.aggregate_evaluate(server_round, test_results, failures)
            test_loss_aggregated, test_metrics_aggregated = test_aggregated_result

            for key, value in test_metrics_aggregated.items():
                val_metrics_aggregated[key] = value
            if test_loss_aggregated is not None:
                val_metrics_aggregated["test - loss - aggregated"] = test_loss_aggregated

        return val_loss_aggregated, val_metrics_aggregated, (results, failures)
```

Comment (emersodb) [reply fl4health/server/base_server.py:224]: This would allow for you to write a test for `_handle_result_aggregation` to ensure that it works as expected even on edge cases etc.

File: tests/server/test_base_server.py [hunk 5]
```diff
@@ -126,7 +126,7 @@ def test_metrics_reporter_fit_round(mock_fit_round: Mock) -> None:
     }
 
 
-@patch("fl4health.server.base_server.Server.evaluate_round")
+@patch("fl4health.server.base_server.FlServer._evaluate_round")
 @freeze_time("2012-12-12 12:12:12")
 def test_metrics_reporter_evaluate_round(mock_evaluate_round: Mock) -> None:
     test_round = 2
```
Comment (emersodb) [inline tests/server/test_base_server.py:132]: I think there are two methods that we want to test in the base server that currently aren't. The first is the functionality of `_unpack_metrics` and the second is `_handle_result_aggregation` from the suggestions above. You should be able to mock up all of the relevant pieces to test those methods out thoroughly using the basic fed avg strategy and our default aggregation schemes in `from fl4health.utils.metric_aggregation`

**Review by emersodb — COMMENTED** [review]
Overall, I think this is a good start. I suggested a few small modifications to the way the code flows to make it a little simpler and easier to test. 

I also suggested to tests that we probably want to add to the base server class to ensure that aggregation is done as expected when the dictionaries are both present or the test components are missing etc.

File: fl4health/clients/basic_client.py (hunk 1 above)
Comment (FreddieNeverLeft) [reply fl4health/clients/basic_client.py:102]: Ok!

File: fl4health/clients/basic_client.py (hunk 2 above)
Comment (FreddieNeverLeft) [reply fl4health/clients/basic_client.py:704]: Was trying to use this to bypass the mypy issues, I will try both with/without as I forgot if this was necessary

Comment (FreddieNeverLeft) [reply fl4health/clients/basic_client.py:704]: test - here is more folloiwing the convention of self.test_metric_manager = MetricManager(metrics=self.metrics, metric_manager_name="test"), maybe rather than adding an enum would it make for sense to use metric_manager_name instead?

File: fl4health/server/base_server.py (hunk 4 above)
Comment (FreddieNeverLeft) [reply fl4health/server/base_server.py:187]: I think this make mroe sense will add the test- related changes here

File: fl4health/server/base_server.py (hunk 3 above)
Comment (FreddieNeverLeft) [reply fl4health/server/base_server.py:224]: similarly, will add the test- related changes here

Comment (FreddieNeverLeft) [reply fl4health/server/base_server.py:224]: will do the testing!

File: tests/server/test_base_server.py (hunk 5 above)
Comment (FreddieNeverLeft) [reply tests/server/test_base_server.py:212]: added the tests in the latest commit!

**Comment by FreddieNeverLeft** [comment]
The changes have been made and the tests have been added!

File: fl4health/clients/basic_client.py [hunk 6]
```diff
@@ -701,8 +701,8 @@ def validate(self) -> Tuple[float, Dict[str, Scalar]]:
             )
             # There will be no clashes due to the naming convention associated with the metric managers
             if self.num_test_samples is not None:
-                val_metrics["test - num_examples"] = int(self.num_test_samples)
-            val_metrics["test - loss"] = test_loss
+                val_metrics[TestMetricPrefix.TEST_PREFIX.value + "num_examples"] = self.num_test_samples
```
Comment (emersodb) [inline fl4health/clients/basic_client.py:704]: I'm good with us defining the constants
```
TEST_NUM_EXAMPLES = f"{TestMetricPrefix.TEST_PREFIX.value} num_examples"
TEST_LOSS = f"{TestMetricPrefix.TEST_PREFIX.value} loss"
```
In the same file that you have the enum and just importing them so you don't have to repeat this structure everywhere.

File: fl4health/server/base_server.py [hunk 7]
```diff
@@ -164,27 +165,71 @@ def poll_clients_for_sample_counts(self, timeout: Optional[float]) -> List[int]:
...
 
-            # Remove loss and num_examples from test_metrics if they exist
-            test_loss = float(test_metrics.pop("test - loss"))
-            test_num_examples = int(test_metrics.pop("test - num_examples"))
+            if len(test_metrics) > 0:
+                assert (
+                    TestMetricPrefix.TEST_PREFIX.value + "loss" in test_metrics
+                    and TestMetricPrefix.TEST_PREFIX.value + "num_examples" in test_metrics
+                ), (
+                    TestMetricPrefix.TEST_PREFIX.value
```
Comment (emersodb) [inline fl4health/server/base_server.py:187]: We use f-strings by convention for this type of thing. If you define the constant I suggested above, this would be
`f"'{TEST_NUM_EXAMPLES}' and '{TEST_LOSS}' keys must be present in test_metrics dictionary for aggregation`

File: tests/server/test_base_server.py [hunk 8]
```diff
@@ -14,11 +17,19 @@
...
 from fl4health.server.base_server import FlServer, FlServerWithCheckpointing
+from fl4health.strategies.basic_fedavg import BasicFedAvg
+from fl4health.utils.metrics import TestMetricPrefix
+from tests.test_utils.custom_client_proxy import CustomClientProxy
 from tests.test_utils.models_for_test import LinearTransform
 
 model = LinearTransform()
 
 
+def metrics_aggregation(to_aggregate: List[Tuple[int, Metrics]]) -> Metrics:
```
Comment (emersodb) [inline tests/server/test_base_server.py:28]: So I know it's a bit of a pain, but let's try using the `evaluate_metrics_aggregation_fn` in `fl4health/utils/metric_aggregation.py` and passing a list with two different EvaluateRes objects. That way we actually see `_handle_result_aggregation` actually extract and computed a weighted average correctly based on num_examples for both a validation and test set.

**Review by emersodb — COMMENTED** [review]
Two final suggestions. One for improving the test and one for reducing a bit of code duplication with the enum.

File: tests/server/test_base_server.py (hunk 8 above)
Comment (FreddieNeverLeft) [reply tests/server/test_base_server.py]: resolved in the latest commit

File: fl4health/server/base_server.py (hunk 7 above)
Comment (FreddieNeverLeft) [reply fl4health/server/base_server.py]: done!

File: fl4health/clients/basic_client.py (hunk 6 above)
Comment (FreddieNeverLeft) [reply fl4health/clients/basic_client.py]: done!

**Comment by FreddieNeverLeft** [comment]
code duplication reduced and test with `evaluate_metrics_aggregation_fn` added!

**Review by emersodb — APPROVED** [review]
Ready to go 🙂
